src/protocol_level_tokens/token.rs

```diff
@@ -0,0 +1,818 @@
+                self.validate_governance_operation(signer.address)?;
+
+                // check if every amount ot be minted have the same decimals as the token.
+                if amounts
+                    .iter()
+                    .any(|amount| amount.decimals() != self.info.token_state.decimals)
+                {
+                    return Err(TokenError::InvalidTokenAmount);
+                }
+            }
+        }
+
+        let amounts = match opts.autoscale {
+            Some(true) => amounts
+                .into_iter()
+                .map(|amount| self.scale_amount(amount))
+                .collect(),
+            Some(false) | None => Ok(amounts),
+        }?;
+
+        let nonce = self
+            .client
+            .get_next_account_sequence_number(&signer.address)
+            .await?
+            .nonce;
+
+        let operations = amounts.into_iter().map(operations::mint_tokens).collect();
+
+        let transaction = send::token_update_operations(
+            &signer,
+            signer.address,
+            nonce,
+            expiry,
+            self.info.token_id.clone(),
+            operations,
+        )?;
+
+        let block_item = BlockItem::AccountTransaction(transaction);
+        Ok(self.client.send_block_item(&block_item).await?)
+    }
+
+    /// Burns a specified amount of tokens.
+    ///
+    /// # Arguments
+    ///
+    /// * `signer` - A [`WalletAccount`] who's address is used as a sender and
+    ///   keys as a signer.
+    /// * `amount` - The amounts of tokens to burn.
+    /// * `expiry` - The optional expiry time for the transaction.
+    /// * `opts` - Options for supply update operations.
+    pub async fn burn(
+        &mut self,
+        signer: &WalletAccount,
+        amounts: Vec<TokenAmount>,
+        expiry: Option<TransactionTime>,
+        opts: Option<SupplyUpdateOptions>,
+    ) -> TokenResult<TransactionHash> {
+        let expiry = expiry.unwrap_or(self.default_expiry);
+        let opts = opts.unwrap_or_default();
+
+        if let Some(validate) = opts.validate {
+            if validate {
+                // check if the signer is authorized to burn tokens.
+                self.validate_governance_operation(signer.address)?;
+
+                // check if every amount ot be burned have the same decimals as the token.
+                if amounts
+                    .iter()
+                    .any(|amount| amount.decimals() != self.info.token_state.decimals)
+                {
+                    return Err(TokenError::InvalidTokenAmount);
+                }
+
+                // check if total amount to burn exceeds total token supply
+                let payload_total = TokenAmount::from_raw(
+                    amounts.iter().fold(0u64, |acc, x| acc + x.value()),
+                    self.info.token_state.decimals,
+                );
+                if self.info.token_state.total_supply.value() < payload_total.value() {
```

Comment:
I assume what we need to check is that the `governance account balance` is smaller than `payload_total` here.